### ISSUE TITLE
Optimization: inference projects the LM head over all 512 positions to use one row

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#206]
**Lane:** partial-cpu — the change and its parity gate are CPU work; only the speedup *number* wants the card.

## What's wrong

`trm/infer.py:91-93` computes the LM head over **all 512 positions** and then throws away 511 of them:

```python
all_logits = run_model_inference(model, padded_tks, depth=depth, new_document=refresh)
logits = all_logits[0, token_idx, :]
```

`token_idx` is a **traced** argument (it is not in `static_argnames`), so this is a dynamic slice and XLA cannot dead-code the unused rows. The full `[1, 512, 50304]` f32 tensor is materialized for every emitted token.

Cost model at the live config (dim 960, vocab 50304, seq 512, 7 encoder blocks + shared refine block):

| | MAC / generated token |
|---|---|
| body (encoder + refine loop + gate), 512 positions | ~86 GMAC (depth 6) |
| **tied LM head, 512 positions** | **~24.7 GMAC** |

That is **~22% of per-token generation compute at depth 6** (~20% at depth 8), plus a **103 MB f32 transient** per token, to produce one 50304-wide row.

## Fix

The `return_hidden` seam already exists (`trm/model/refiner.py:276-282`, `trm/model/refiner_lm.py:72`). Thread the wanted position down and slice `z` **before** the tied-head matmul, so the head projects `[1, 1, dim]` instead of `[1, 512, dim]`.

## Relationship to #153

Independent, and much smaller. #153 (KV cache) makes the forward process one position, which subsumes this — but #153 is a multi-day rewrite and this is a handful of lines that lands on its own. If #153 lands first, close this as `superseded-by #153`.

## Acceptance

- **Correctness (CPU, blocking):** greedy generation from a fixed prompt + seed under `FORCE_F32_COMPUTE` produces the **identical token sequence** before and after. Slicing before vs after a matmul is exact, not approximate — anything short of an exact match is a bug. Goes in `tests/core/`.
- **Speed (GPU, follow-up):** tokens/sec on a 256-token continuation, before vs after. Expect ~1.25×; report the actual number, no claim without it.

**Comment by MatthewLacerda2** [comment]
**The pre-registered acceptance bar is not achievable, and the reason is worth recording.**

#206 says: *"Slicing before vs after a matmul is exact, not approximate — anything short of an exact match is a bug."* That premise is false in floating point.

Measured by feeding the **same** pre-head state to both shapes, so nothing but the matmul differs:

```
matmul [1, 512, 32] @ [32, 37]   vs   matmul [1, 1, 32] @ [32, 37]
  bit-identical : False
  max abs diff  : 2.384e-07
  max rel diff  : 1.816e-05
  argmax agrees : True
```

Float matmul is **not shape-invariant**. XLA selects a different kernel and reduction order for a 512-row operand than for a 1-row one, so the last ulp moves. The identity is exact in real arithmetic and inexact in f32, and no amount of correct slicing changes it. This is the same class as the ~1e-7 golden-run drift already known on this box.

## What the gate became instead

Two conditions rather than one:

1. **Agreement to a tolerance far below anything meaningful** — `rtol=1e-4, atol=1e-5`, against a val-CE noise floor of σ≈0.03 nats, roughly five orders of magnitude larger.
2. **Exact agreement on the argmax**, plus exact agreement on which tokens survive top-k/top-p truncation.

Condition 2 is the one that carries the weight. Generation reads exactly one thing out of these numbers — which token wins. A logit that drifts in its last ulp without changing the decision is not a behaviour change; a drifting argmax would be, and that is now pinned directly.

## The honest residual risk

It is **not zero**. If the top two logits ever sit within ~2.4e-7 of each other, the sliced and full paths can pick different tokens, and from there the continuations diverge completely. At the live config (max|logit| ~21, typical top-2 gaps of order 0.1–1) this is rare, but it is not impossible, and I would rather it be written down than discovered.

Practical consequence for #203's transcript logbook: entries generated before and after this commit are not guaranteed token-identical. That is exactly what the `tool_commit` field in the frontmatter is for — reproducibility holds *within* a code version, which is the property the logbook actually needs.

Proceeding on the judgement that ~22% of per-token generation compute is worth 1e-7 of logit drift that provably does not move any decision in test. Flagging rather than burying it, since the bar was pre-registered.
